6661: Improve timezone handling documentation

heiglandreas opened the pull request:
This commit alters the page to better handle datetimes with timezones. The result will store the datetime with the local date as datetime-value and the timezone as string-value. That way the database-specific ways of doing datetime arithmentics with timezones (see examples for [MySQL](https://andreas.heigl.org/2016/04/18/timezones-and-mysql/) and [Postgres](https://andreas.heigl.org/2016/05/29/timezones-and-postgresql/)) are still possible, the datetime will not be prone to errors due to timezone-changes between creation and the actual datetime and still the application will be able to use a generic datetime-object that can then be converted to a local timezone of the current user.

heiglandreas committed:
Alters the examples to better handle timezones

This commit alters the page to better handle datetimes with timezones. The result will store the datetime with the local date as datetime-value and the timezone as string-value. That way the database-specific ways of doing datetime arithmentics with timezones (see examples for [MySQL](https://andreas.heigl.org/2016/04/18/timezones-and-mysql/) and [Postgres](https://andreas.heigl.org/2016/05/29/timezones-and-postgresql/)) are still possible, the datetime will not be prone to errors due to timezone-changes between creation and the actual datetime and still the application will be able to use a generic datetime-object that can then be converted to a local timezone of the current user.

heiglandreas commented:
This PR would fix #6661

Ocramius commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -177,7 +113,12 @@ requiring timezoned datetimes:
             if (!$this->localized) {
-                $this->created->setTimeZone(new \DateTimeZone($this->timezone));
+                
+                $class = $this->created::class;
```
`::class` is not an actual constant, and will not resolve at runtime.

Ocramius commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -177,7 +113,12 @@ requiring timezoned datetimes:
-                $this->created->setTimeZone(new \DateTimeZone($this->timezone));
+                
+                $class = $this->created::class;
+                $this->created = new $class(
```
I suggest using one of either `DateTimeImmutable` or `DateTime`. Since DBAL 2.6, `DateTimeImmutable` is viable, so I suggest just going with that rather than with a generic example

Ocramius commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -67,84 +67,20 @@ The problem is simple. Not a single database vendor saves the timezone, only the
+The solution for this dilemma seems simple. Don't use timezones with DateTime and Doctrine 2. However there is a workaround
 that even allows correct date-time handling with timezones:
 
-1. Always convert any DateTime instance to UTC.
```
I suggest adding a section that links to your article about the original problem. Most people will try looking for a simple "convert to UTC" solution, and they will mess it up.

Ocramius reviewed (changes requested):
Overall good improvement, but IMO the `UTC` approach should be perfectly viable for logging events (specifically in the past, while the issue with zone time changes is for future data).

A typical requirement is to be able to sort data based on a `DATETIME` field: could you think about how to integrate that into the article?

heiglandreas committed:
Updated proposed  change for timezonehandling

This commit addresses the concerns @ocamius raised and reintroduces the section about converting datetimes to UTC along with a short description of when to use which approach along with links to different blogposts.

lcobucci commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -58,25 +58,88 @@ Handling different Timezones with the DateTime Type
 
-The solution for this dilemma is simple. Don't use timezones with DateTime and Doctrine 2. However there is a workaround
-that even allows correct date-time handling with timezones:
+The solution for this dilemma seems simple. Don't use timezones with DateTime and Doctrine 2. However there are 2 workarounds
```
> The solution for this dilemma seems simple. Don't use timezones with DateTime and Doctrine 2.

Maybe is better?:

The solution for this dilemma seems simple: don't use timezones with DateTime and Doctrine 2.

lcobucci commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -58,25 +58,88 @@ Handling different Timezones with the DateTime Type
 
-1. Always convert any DateTime instance to UTC.
-2. Only set Timezones for displaying purposes
+1a. Don't convert DateTimes to UTC
```
Not sure if it's me, but this is a bit confusing... we mention "2 workarounds" and then present a list with 3 (4) items but these workarounds are actually explained as sub-sections of that area. 😕

heiglandreas replied:
I've modified that part

lcobucci commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -58,25 +58,88 @@ Handling different Timezones with the DateTime Type
+
+Say we have an international calendaring application where users can add events that occur at different places worldwide and therefore at different timezones. To determine the exact time an event will happen means to save both the local time of the event and the timezone the event will happen in.
+
+Using the defalt DateTimeType will store the DateTime in the local time but without the timezone information. Therefore we need to store the timezone information as well and also need to provide a way to get the datetime back from the database with the correct timezone-information.
```
Typo: `defalt` -> `default`

heiglandreas replied:
Thanks for noting. I've changed that.

lcobucci commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -58,25 +58,88 @@ Handling different Timezones with the DateTime Type
+        {
+            $this->localized = true;
+            $this->eventDateTime = $eventDateTime;
+            $this->timezone = $$eventDateTime->getTimeZone()->getName();
```
Typo `$$eventDateTime` -> `$eventDateTime`

lcobucci reviewed (changes requested):
It's going on good direction IMO, just some fixes/comments 😄

Majkl578 commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -58,25 +58,88 @@ Handling different Timezones with the DateTime Type
-The solution for this dilemma is simple. Don't use timezones with DateTime and Doctrine 2. However there is a workaround
-that even allows correct date-time handling with timezones:
+The solution for this dilemma seems simple. Don't use timezones with DateTime and Doctrine 2. However there are 2 workarounds
+that even allows correct date-time handling with timezones depending on your use case:
```
_that allow_

heiglandreas replied:
Thanks for noting. I've modified that part accordingly

Majkl578 commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -58,25 +58,88 @@ Handling different Timezones with the DateTime Type
+    /**
+     * @Entity
+     */
+    class Event
```
Strictly speaking, this is not a valid entity since it has no identifier, can we add one (or make one field one)?

heiglandreas replied:
I was copy/pasting the example beginning at line 16 here which also does not have an identifier. Shall I add one everywhere? Or shall we skip that here as it is not relevant to the topic illustrated here?

SenseException replied:
I suggest to add them in this file. In some past PRs we also added some missing lines and the identifier doesn't take much space that could distract from the core information. Since there are namespaces in use, some of the annotations need to be included with `use` too

heiglandreas replied:
This should be addressed now

Majkl578 commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -58,25 +58,88 @@ Handling different Timezones with the DateTime Type
+        {
+            if (!$this->localized) {
+                
+                $this->eventDateTime = new DateTimeImmutable(
```
This will trigger entity update as a side effect, that is not desired.

heiglandreas replied:
How can that be avoided? I'm not too deep into that, so any help would be appreciated!

SenseException replied:
What is the method supposed to do? Its name suggests just to return a datetime value but it changes `$eventDateTime` property.

heiglandreas replied:
Added the reasoning in a code-comment.

I'd rather remove that but that would mean that there needs to be a way to modify that value correctly during hydration which currently doesn't work as I can only access the current value while converting from a DB value to a PHP value. But I would also need access to all the other values to be able to get the timezone string :frowning:

SenseException replied:
Instead of writing the datetime back into the `$eventDateTime`, it could be returned directly. It doesn't have to return the same instance, does it? This way the update trigger can be prevented.

Or instead of having a boolean `$localized` property maybe you can store the created DateTimeImmutable in that property (With a different name) if a same instance matters.

What do you think about putting the comment below this class example? It seems like it belongs to the documentation itself than into a comment.

Majkl578 reviewed (changes requested):
I don't particulary like the naming inconsistencies in the text - once it's _DateTime_, then _datetime_, then _date-time_. Also we have DateTimeImmutable for a while which is actually preferred nowadays. Can we sync it to one variant?

lcobucci commented:
I've removed this issue from `2.6` since we can update the docs anytime

landsman commented:
This would be great!

heiglandreas committed:
Adapt reference to nore than 2 possibilities

That was mentioned during the CodeReview by Majkl578 and lcobucci

heiglandreas committed:
Unify spelling of datetime and DateTime

heiglandreas committed:
Fix bug caused by double "$" instead of one

heiglandreas committed:
Use DateTimeImmutable as default

As DateTimeImmutable is what DateTime should have been in the first
place let's use that at least for the docs

heiglandreas commented:
Hey folks: Can we get on with this somehow? There are some questions unanswered for some time now. So are these still a blocker? Or do we just close the PR and go on?

I'd rather not see the last option as the currently available information on the docs are actually harming users as the proposed solution will sooner or later harm users by introducing situations where wrong data is stored.

SenseException commented:
@heiglandreas The branches changed between the last changes and the new default branch approach. Can you please change the base branch of `heiglandreas:patch-1` to `2.9.x`? Then let's take a fresh look into this PR.

heiglandreas committed:
Adapt to new base-branch

heiglandreas committed:
Address reasoning for changing value in accessor

SenseException commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -67,53 +68,151 @@ The problem is simple. Not a single database vendor saves the timezone, only the
 
-The solution for this dilemma is simple. Don't use timezones with DateTime and Doctrine ORM. However there is a workaround
-that even allows correct date-time handling with timezones:
+The solution for this dilemma seems simple: don't use timezones with DateTime(Immutable) and Doctrine 2. However there are some
```
```suggestion
The solution for this dilemma seems simple: don't use timezones with DateTime(Immutable) and Doctrine ORM. However there are some
```

SenseException commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -67,53 +68,151 @@ The problem is simple. Not a single database vendor saves the timezone, only the
+    }
+
+Using this way of handling timezones allows you also to use the database-specific ways of
+doing DateTime-arithmetics with the appropriate timezones. Make sure though that the database
```
People might confuse that the timezone-database is related to ORM.

```suggestion
doing DateTime-arithmetics in SQL with the appropriate timezones. Make sure though that the database
```

SenseException reviewed (changes requested):
Thank you for coming back to this PR. :-)

heiglandreas committed:
Implement changes from Code-Review

SenseException commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -67,53 +68,190 @@ The problem is simple. Not a single database vendor saves the timezone, only the
+            $this->timezone = $eventDateTime->getTimeZone()->getName();
+        }
+
+        /** @PostLoad */
```
```suggestion
        /** @ORM\PostLoad */
```

SenseException commented on docs/en/cookbook/working-with-datetime.rst:
```diff
@@ -67,53 +68,190 @@ The problem is simple. Not a single database vendor saves the timezone, only the
+        /** @PostLoad */
+        public function correctTimezone(): void
+        {
+            $this->eventDateTime = new DateTimeImmutable(
```
Writing a new DateTimeImmutable into `$this->eventDateTime` will trigger an update for the entities when `$entityManager->flush()` is called. Do you need this alternative `Event` entity example?

heiglandreas replied:
As this function is done via a `PostLoad`-Action I assumed that it does **not** trigger the update during the `flush`. But I then might have missunderstood [the docs regarding lifecycle-callbacks]( https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/reference/events.html#lifecycle-callbacks). Espechially as the hydrated content is indeed not the one  the class was initially created with as there seems to be no way to actually get the `eventDateTime` property into the initial state without the `PostLoad` action.

Any ideas how to solve that dilemma?

heiglandreas replied:
I just did a quick check. As far as I understand [the code](https://github.com/doctrine/orm/blob/2.9.x/lib/Doctrine/ORM/UnitOfWork.php#L2941), the `PostLoad` event listeners are run as the last part of the hydration process. Therefore I expect that the entity is not marked as changed as the `originalEntityData` array is populated with the data **after** the hydration finished completely. And therefore contained the already changed Data.

So that process should indeed be the preffered one. Or I completely misunderstood the code...

SenseException replied:
Yes, having the following code in an Entity

``` php
    /** @ORM\PostLoad */
    public function fooLoad(): void
    {
        $this->bar = 'lol';
    }
```

and running 


``` php
$foo = $this->entityManager->find(\App\Entity\Foo::class, 1);
$this->entityManager->flush();
```

Will write "lol" into the mapped table column of `$this->bar`.

---

There's also a note about combined usage of `PostLoad` and `Doctrine\ORM\AbstractQuery#toIterable()` that is not considered safe.
https://www.doctrine-project.org/projects/doctrine-orm/en/2.9/reference/events.html#lifecycle-events

heiglandreas committed:
Fix missing Namespace-prefix

Co-authored-by: Claudio Zizza <859964+SenseException@users.noreply.github.com>

heiglandreas committed:
Fix weird UnitOfWork-behaviour

SenseException commented on tests/Doctrine/Tests/Examples/DateTimeExampleTest.php:
```diff
@@ -0,0 +1,54 @@
+        $schemaTool->createSchema([$classes]);
+    }
+
+    public function testDateTimeWithTimezoneExample(): void
```
Is the test currently a work in progress?

heiglandreas commented:
No worries @SenseException: Life just got in the way. And I'm currently trying to wrap my head around the Embeddable idea.

SenseException commented:
Thanks for the update. Feel free to catch on when you're ready. :-)
